Написала автотесты для задания Спринт_5

Elena120393 opened the pull request.

Elena120393 committed:
Написала автотесты для задания Спринт_5

AlexKitIt commented on locators.py:
```diff
@@ -0,0 +1,67 @@
+from selenium.webdriver.common.by import By
+
+
+BASE_URL = "https://stellarburgers.nomoreparties.site"  #Базовый URL сайта
+
+# Учетные данные для входа и регистрации
+VALID_NAME = "EalenaAlena"
+# Для избежания повторной регистрации генерируем уникальный email
+VALID_EMAIL = "ea78elena_artemeva_19_789@gmail.com"
+VALID_PASSWORD = "aelena_artemeva_19_789@gmail.com"  # Корректный пароль, длиной >= 6 символов
```
Нужно исправить: урлы и тестовые данные не имеют отношения к локаторам, их необходимо перенести в модуль data

AlexKitIt commented on locators.py:
```diff
@@ -0,0 +1,67 @@
+# Локаторы для страницы регистрации
+NAME_FIELD = (By.XPATH, "//*[@id='root']/div/main/div/form/fieldset[1]/div/div/input")
+EMAIL_FIELD = (By.XPATH, "//*[@id='root']/div/main/div/form/fieldset[2]/div/div/input")
+PASSWORD_FIELD = (By.NAME, "Пароль")
```
Нужно исправить: здесь и далее: использование порядкового номера элемента в dom является весьма нестабильным способом поиска, необходимо привязываться к атрибутам конкретного элемента
Кроме того ходить от root и использовать абсолютные пути нельзя - очень ненадежно, добавится на пути новый элемент - тест ляжет. Все подобные локаторы необходимо переписать

AlexKitIt commented on tests/test_constructor_navigation_tabs.py:
```diff
@@ -0,0 +1,51 @@
+
+
+
+def test_constructor_navigation_tabs(driver):
```
Нужно исправить: здесь и далее по тестам: тестовые методы в модулях объединяются в тестовые классы, необходимо создать в каждом тестовом модуле тестовые классы и разместить в них методы, даже если тестовый метод только один

AlexKitIt commented on tests/test_constructor_navigation_tabs.py:
```diff
@@ -0,0 +1,51 @@
+    # Вкладка "Булки" - проверка наличия кнопки (изображения) для булки
+    buns_element = WebDriverWait(driver, 10).until(EC.element_to_be_clickable(BUNS_TAB))
+    buns_element.click()
+    time.sleep(3)  # Ждем обновления состояния вкладки
```
Нужно исправить: здесь и далее: слипов и неявных ожиданий быть не должно. Необходимо использовать expected_conditions и встраивать ожидания на нем.

AlexKitIt commented on tests/test_constructor_navigation_tabs.py:
```diff
@@ -0,0 +1,51 @@
+
+    # Проверяем, что на странице присутствует элемент-кнопка с изображением булки
+    buns_img = driver.find_element(By.XPATH,
+                                   "//img[@src='https://code.s3.yandex.net/react/code/bun-01.png' and @alt='Флюоресцентная булка R2-D3']")
```
Нужно исправить: здесь и далее: все локаторы хранятся отдельно от тестов в модуле locators, а тут только переменные, в которые они были записаны

AlexKitIt commented on tests/test_constructor_navigation_tabs.py:
```diff
@@ -0,0 +1,51 @@
+    # Проверяем, что на странице присутствует элемент-кнопка с изображением булки
+    buns_img = driver.find_element(By.XPATH,
+                                   "//img[@src='https://code.s3.yandex.net/react/code/bun-01.png' and @alt='Флюоресцентная булка R2-D3']")
+    assert buns_img.is_displayed(), "Кнопка для булки не отображается, несмотря на активную вкладку 'Булки'"
```
Нужно исправить: здесь и далее по тестам табов: все эти элементы есть с самого начала прогрузки страницы, проверять по ним корректность переходов нельзя. Проверять стоит по новому классу элемента после клика

AlexKitIt replied:
Нужно исправить: тестов на переход по табам должно быть три отдельных, по одному на каждый таб

AlexKitIt commented on tests/test_login_from.py:
```diff
@@ -0,0 +1,154 @@
+    assert order_button is not None, "Кнопка 'Оформить заказ' не найдена на главной странице после входа"
+
+    # Закрытие драйвера
+    driver.quit()
```
Нужно исправить: здесь и далее: инстанс и так закрывается в фикстуре

AlexKitIt commented on tests/test_registration.py:
```diff
@@ -0,0 +1,50 @@
+    driver.get(f"{BASE_URL}/register")
+    WebDriverWait(driver, 10).until(EC.presence_of_element_located(NAME_FIELD))
+    driver.find_element(*NAME_FIELD).send_keys('Alenka')
+    driver.find_element(*EMAIL_FIELD).send_keys('Alenkaart_artemeva_19_789@gmail.com')
```
Нужно исправить: для теста регистрации email должен генерироваться каждый раз новый, в противном случае мы получим ошибку уже созданного пользователя. Необходимо написать метод, генерирующий email и положить его в helpers

AlexKitIt commented on tests/test_registration.py:
```diff
@@ -0,0 +1,50 @@
+         - Пароль, длиной не менее 6 символов.
+    3. После нажатия на "Зарегистрироваться" ожидается появление индикатора успешной регистрации (например, "Личный кабинет").
+    """
+    driver.get(f"{BASE_URL}/register")
```
Можно лучше: эндпоинты стоит хранить отдельно, они постоянно переиспользуются, для них стоит создать модуль urls или отдельный класс в data

Elena120393 committed:
Обогатила файл с локаторами.
Были созданы файлы urls, helpers, test_data.
Были обновлены кейсы в соответствии с замечаниями.

AlexKitIt commented on helpers.py:
```diff
@@ -0,0 +1,15 @@
+    unique_str = uuid.uuid4().hex[:8]
+    return f"test_{unique_str}@{domain}"
+
+if __name__ == "__main__":
+    # Пример использования
+    print(generate_unique_email("gmail.com"))
```
Нужно исправить: отладочной информации быть в проекте не должно

AlexKitIt commented on locators.py:
```diff
@@ -0,0 +1,73 @@
+
+
+# Локаторы для страницы регистрации
+NAME_FIELD = (By.XPATH, "//*[@id='root']/div/main/div/form/fieldset[1]/div/div/input")
```
Повторный коммент: здесь и далее: ВСЕ локаторы с абсолютным путем, индексами, с указанием @id='root' в пути должны быть изменены

AlexKitIt commented on tests/test_constructor_navigation_tabs.py:
```diff
@@ -0,0 +1,32 @@
+        driver.find_element(*BUNS_TAB).click()
+        # Проверка, что мы в разделе "Булки"
+        buns_page = driver.find_element(*BUNS_TAB)
+        assert buns_page.is_displayed(), "Не удалось перейти в раздел Булки."
```
Нужно исправить: здесь и далее по тестам табов: все эти элементы есть с самого начала прогрузки страницы, проверять по ним корректность переходов нельзя. Проверять стоит по новому классу элемента после клика

AlexKitIt commented on tests/test_login_from.py:
```diff
@@ -0,0 +1,139 @@
+)
+
+# Константы, определяющие валидные данные для логина
+VALID_EMAIL = "elena_artemeva_19_789@gmail.com"
+VALID_PASSWORD = "elena_artemeva_19_789@gmail.com"
```
Нужно исправить: это уже вынесено в data, тут быть не должно

AlexKitIt commented on tests/test_login_from.py:
```diff
@@ -0,0 +1,139 @@
+VALID_EMAIL = "elena_artemeva_19_789@gmail.com"
+VALID_PASSWORD = "elena_artemeva_19_789@gmail.com"
+
+PASSWORD_RECOVERY_LOGIN_LINK = (By.XPATH, "//a[text()='Войти']")
```
Нужно исправить: все локаторы уже в locators, тут быть не должно

AlexKitIt commented on tests/test_registration.py:
```diff
@@ -0,0 +1,70 @@
+from helpers import generate_unique_email
+
+
+def test_registration_success(driver):
```
Повторный коммент: Нужно исправить: здесь и далее по тестам: тестовые методы в модулях объединяются в тестовые классы, необходимо создать в каждом тестовом модуле тестовые классы и разместить в них методы, даже если тестовый метод только один

Elena120393 committed:
Обогатила файл с локаторами.
Были созданы файлы urls, helpers, test_data.
Были обновлены кейсы в соответствии с замечаниями.

AlexKitIt commented on tests/test_constructor_navigation_tabs.py:
```diff
@@ -0,0 +1,66 @@
+        buns_tab.click()
+
+        # Находим родительский элемент (используя xpath './..')
+        parent_element = buns_tab.find_element(By.XPATH, "./..")
```
Нужно исправить: здесь и далее: все локаторы хранятся отдельно от тестов в модуле locators, а тут только переменные, в которые они были записаны

AlexKitIt commented on tests/test_login_from.py:
```diff
@@ -0,0 +1,144 @@
+)
+
+# Константы, определяющие валидные данные для логина
+VALID_EMAIL = "elena_artemeva_19_789@gmail.com"
+VALID_PASSWORD = "elena_artemeva_19_789@gmail.com"
```
Повторный коммент: Нужно исправить: данные уже вынесены в data, в тестах модуля должны использоваться они, поддержка должна осуществляться из одного места. Тестовые модули не предназначены для хранения тестовых данных

AlexKitIt commented on tests/test_login_from.py:
```diff
@@ -0,0 +1,144 @@
+VALID_EMAIL = "elena_artemeva_19_789@gmail.com"
+VALID_PASSWORD = "elena_artemeva_19_789@gmail.com"
+
+PASSWORD_RECOVERY_LOGIN_LINK = (By.XPATH, "//a[text()='Войти']")
```
Повторный коммент: Нужно исправить: локаторы хранятся в locators. Этот оставлен здесь, так быть не должно

AlexKitIt commented on tests/test_login_from.py:
```diff
@@ -0,0 +1,144 @@
+        driver.find_element(*LOGIN_EMAIL_FIELD).send_keys(VALID_EMAIL)
+        driver.find_element(*LOGIN_PASSWORD_FIELD).send_keys(VALID_PASSWORD)
+
+        time.sleep(7)
```
Нужно исправить: здесь и далее: слипов и неявных ожиданий быть не должно. Необходимо использовать expected_conditions и встраивать ожидания на нем.

Elena120393 committed:
Внесла исправления:
переписала локатор, где использовался порядковый номер элемента в dom;
добавила локатор и переписала кейс с конструктором;
удалила слипы

AlexKitIt commented on tests/test_login_from.py:
```diff
@@ -0,0 +1,140 @@
+)
+
+# Константы, определяющие валидные данные для логина
+VALID_EMAIL = "elena_artemeva_19_789@gmail.com"
+VALID_PASSWORD = "elena_artemeva_19_789@gmail.com"
```
Повторный коммент: Нужно исправить: данные уже вынесены в data, в тестах модуля должны использоваться они, поддержка должна осуществляться из одного места. Тестовые модули не предназначены для хранения тестовых данных
